idseq_dag/util/s3.py

```diff
@@ -140,9 +152,22 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
         # This check is a bit imperfect when untarring... unless you follow the discipline that
         # all contents of file foo.tar are under directory foo/...
         if os.path.exists(dst):
-            # No need to fetch this file from s3, it has been just produced
-            # on this instance.
-            return dst
+            # Destination filename exists. If it is in the reference download directory, check the reference fetch log
+            # for a record of the source key and etag, and short-circuit the download if they match. Otherwise, blindly
+            # assume that it's the same file and short-circuit the download.
+            if abspath.startswith(config["REF_DIR"]):
+                try:
+                    with open(os.path.join(config["REF_FETCH_LOG_DIR"], abspath_hash)) as fh:
+                        fetch_record = json.load(fh)
+                    obj = s3.Bucket(parsed_s3_url.netloc).Object(parsed_s3_url.path.lstrip('/'))
```

Comment:
Does this fail if the object doesn't exist in s3?  Or is it just hashing just the URL?

I ask because in parallel with you I was trying to cleanup and reduce calls to s3 from the wrapper function around this whole function, which tries to guess the correct compression extension and checks repeatedly if various compressed versions of the file exist.   I wanted to make sure that check only happens if the file actually needs to be downloaded. m

Reply:
I think if I understand this design correctly, it definitely polls S3, which is good for ensuring that it always fetches the newest file (if it got updated).   I think this is more valuable than trying to reduce S3 calls.  Thanks.

Reply:
Back to David's concern about locking.   I think, unfortunately, locks[abspath] aka destination_lock is still a threading.RLock, from the days before we used multiprocessing.  It should be a multiprocessing.RLock now for extra safety.  Same with mutex.  

I'll also restate my question - is it safe to use the same global boto3 session from multiple threads and from multiple processes?   If the answer is no, it would rather easy to solve the problem here by creating a separate session for each destination.  The multiple threads/processes serialize by destination --- this is what the locks in this design ensure.  A session per call would be fine too, for fetch_from_s3.  I am more concerned about that with check_s3_presence.

Reply:
There is some uncertainty about how safe it is to use the same boto3 client instance from multiple threads, so I'm going to make each subroutine spawn its own client.

Reply:
Thank you.